Add max crawl size option to backend and frontend

tw4l opened the pull request:
Fixes #695  

Includes backend nightly test.

## Screenshots

Crawl config editor:

<img width="910" alt="Screen Shot 2023-08-03 at 3 04 42 AM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/a05d89d9-03c7-4739-90e8-ef7495077267">

Config details:

<img width="175" alt="Screen Shot 2023-08-03 at 3 09 32 AM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/5bdcb194-5dab-450c-8a66-5b1819869e8e">

tw4l committed:
Add sizeLimit to backend

tw4l committed:
Add size limit in GB to frontend

tw4l committed:
Fix workflow editor help text

tw4l committed:
Add Crawl Size Limit to workflow setup docs

tw4l committed:
Change config label from Size Limit to Crawl Size Limit

tw4l committed:
Fix revs history test to account for changes

tw4l commented:
Currently works via check in operator during crawl sync, but I think the `sizeLimit` option should work on its own in the crawler without needing the extra operator check so perhaps something is off.

tw4l commented:
Should rename this field, as `sizeLimit` is already being used in the crawler to rotate the crawler pods and commit any data written so far. When a crawler reaches current sizeLimit, we commit that data to WACZs, upload and restart the crawler with an empty directory, which keeps the WACZs to a reasonable size, and also we do this by time.

So for the check in the operator as implemented here, we should likely use a different name - perhaps `maxCrawlSize`.

tw4l committed:
Rename setting to maxCrawlSize

tw4l committed:
Rename to maxCrawlSize in frontend

tw4l committed:
Remove MB fallback in config details since editor only allows GB

tw4l committed:
Rename test function

tw4l committed:
Add comment to test fixture

tw4l commented:
Testing locally now and then will make this ready for review again.

tw4l commented:
This is tested after the changes and ready for review.

One note, we may want to consider setting `operator_resync_seconds` to default lower than 10 seconds. In testing, with a max crawl size of 1 GB and the default operator_resync_seconds set at 10, the crawl tests to finish at around 1.3 - 1.5 GB for a large site. Of course part of this is simply finishing the already pending URLs, but the sooner we send the stopping signal the fewer URLs we'll end up with in the WACZ.

tw4l commented:
Rebased on latest main for test fixes

SuaYoo commented on frontend/src/components/config-details.ts:
```diff
@@ -96,6 +96,28 @@ export class ConfigDetails extends LiteElement {
       }
     };
 
+    const renderSize = (valueBytes?: number | null, fallbackValue?: number) => {
```
We use the shoelace [`sl-format-bytes`](https://shoelace.style/components/format-bytes/) component to render bytes, should work in this case, too!

```ts
<sl-format-bytes
    value=${valueBytes}
    display="narrow"
></sl-format-bytes>
```

tw4l replied:
I tried using this but `sl-format-bytes` uses 1000 bytes/kb instead of 1024 bytes/kb, which results in the config detail having an inaccurate number (e.g. "3.22 GB" instead of "3 GB"). It might be better to keep it as-is but use the same formatting for the GB suffix as `<sl-format-bytes display="narrow">` - what do you think?

tw4l replied:
Or we could make a custom `<btrix-format-bytes>` that calculates with 1024 bytes/kb?

SuaYoo replied:
Got it! It looks like `sl-format-bytes` uses [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat#syntax) under the hood, so to format it consistently for different locales you would do something like:
```ts
const value = 1000; // example
const formatter = new Intl.NumberFormat(undefined, {
  style: "unit",
  unit: "gigabyte",
  unitDisplay: "narrow"
});

formatter.format(1000); // outputs "1,000GB"
```

So lines 102-118 could look something like (untested!):
```ts
const formatter = new Intl.NumberFormat(undefined, {
  style: "unit",
  unit: "gigabyte",
  unitDisplay: "narrow"
});
if (valueBytes) {
  const sizeGB = Math.floor(valueBytes / bytesPerGB);
  return formatter.format(sizeGB);
}
if (typeof fallbackValue === "number") {
  let value = "";
  if (fallbackValue === Infinity) {
    value = msg("Unlimited");
  } else if (fallbackValue === 0) {
    value = formatter.format(0);
  } else {
    value = formatter.format(Math.round(fallbackValue / bytesPerGB));
  }
  return html`<span class="text-neutral-400"
    >${value} ${msg("(default)")}</span
  >`;
}
```

tw4l replied:
Change pushed and tested, thanks for digging on this!

SuaYoo commented on frontend/src/pages/org/workflow-editor.ts:
```diff
@@ -487,6 +489,13 @@ export class CrawlConfigEditor extends LiteElement {
     };
 
+    const bytesToGB = (value: any, fallback: number | null) => {
+      const bytesPerGB = 1073741824;
```
Could store this constant outside of the class definition, around line 216 like `DEFAULT_BEHAVIORS`, to use on line 2139

tw4l replied:
Updated

tw4l committed:
Use BYTES_PER_GB as constant across component

tw4l committed:
Use Intl.NumberFormat to format GBs

tw4l commented:
@SuaYoo ready for another look

ikreymer commented on backend/btrixcloud/operator.py:
```diff
@@ -624,6 +626,12 @@ async def update_crawl_state(self, redis, crawl, status):
                 + "gracefully stopping crawl"
             )
 
+        if crawl.max_crawl_size and stats["size"] > crawl.max_crawl_size:
```
```suggestion
        if crawl.max_crawl_size and status.size and status.size > crawl.max_crawl_size:
```
Just in case -- stats["size"] may not exist and already updating `status.size`

ikreymer replied:
Maybe should move this to be below / as part of this conditional?
```
        if stats["size"] is not None:
            status.size = humanize.naturalsize(stats["size"])
```

tw4l replied:
I moved this up a bit, since we want to set stopping earlier. Go ahead and take a look.

tw4l replied:
Hm, this seems to have broken some tests, I'll take a look in a bit

tw4l replied:
Yeah seems better now. Defaulting back to status.filesAddedSize if stats["size"] isn't set

ikreymer replied:
Yep, I think that makes sense - the filesAddedSize should always be <= current size, so that makes sense as a fallback.

tw4l committed:
Use status.size

tw4l committed:
Have another go at fixing size

ikreymer commented on backend/btrixcloud/operator.py:
```diff
@@ -210,6 +211,7 @@ async def sync_crawls(self, data: MCSyncData):
             started=data.parent["metadata"]["creationTimestamp"],
             stopping=spec.get("stopping", False),
             expire_time=from_k8s_date(spec.get("expireTime")),
+            max_crawl_size=configmap["MAX_CRAWL_SIZE"],
```
```suggestion
            max_crawl_size=configmap.get("MAX_CRAWL_SIZE", 0)
```

It's possible for this not to exist on old older configmaps that were created before this property is set.

ikreymer committed:
convert configmap to use get for MAX_CRAWL_SIZE

ikreymer committed:
lint fix

ikreymer committed:
operator cleanup:
- use 'status.stopping' instead of 'crawl.stopping' to indicate crawl is being stopped, as changing later has no effect in operator
- add simple is_crawl_stopping() to return if crawl is being stopped, based on crawl.stopping or size or time limit being reached
- crawlerjob status: store byte size under 'size', human readable size under 'sizeHuman' for clarity
- size stat always exists so remove unneeded conditional (defaults to 0)
- subchart: update crawlerjob crd in btrix-crds to show status.stopping and also show size via sizeHuman field,
bump subchart version to 0.1.1

ikreymer committed:
update chart

ikreymer committed:
add missing btrix-crds-0.1.1.tgz

SuaYoo reviewed (approved):
Haven't tested the backend changes locally, frontend changes look good

ikreymer commented on frontend/src/pages/org/workflow-editor.ts:
```diff
@@ -213,6 +215,7 @@ const DEFAULT_BEHAVIORS = [
   "autofetch",
   "siteSpecific",
 ];
+const BYTES_PER_GB = 1073741824;
```
We should decide if we want to do GB or GiB? I wonder if GiB might be more user-friendly, but either way can do a consistency pass on this later.

ikreymer commented:
Should be good to go - started working on a good commit message to cover all the final work
